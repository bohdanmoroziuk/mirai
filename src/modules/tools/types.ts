export interface Group {
  id: string;
  name: string;
}

export type GroupId = Group['id'];

export type GroupName = Group['name'];

export interface Tool {
  id: string;
  name: string;
  url: string;
  image: string;
  group: Group | null;
}

export type ToolId = Tool['id'];

export type ToolBody = Omit<Tool, 'id'>;
