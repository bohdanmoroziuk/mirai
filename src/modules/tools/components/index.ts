export { default as ToolCard } from 'src/modules/tools/components/ToolCard.vue';
