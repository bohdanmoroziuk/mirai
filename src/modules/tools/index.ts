export * from 'src/modules/tools/types';
export * from 'src/modules/tools/routes';
export * from 'src/modules/tools/stores';
export * from 'src/modules/tools/services';
export * from 'src/modules/tools/components';
