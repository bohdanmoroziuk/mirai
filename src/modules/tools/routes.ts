import { RouteRecordRaw } from 'vue-router';

export const routes: RouteRecordRaw[] = [
  {
    path: 'tools',
    component: () => import('src/modules/tools/pages/Index.vue'),
    children: [
      {
        path: '',
        name: 'tools',
        component: () => import('src/modules/tools/pages/Tools.vue'),
      },
    ],
  },
];
