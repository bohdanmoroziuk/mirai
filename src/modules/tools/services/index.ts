export * from 'src/modules/tools/services/tools';
