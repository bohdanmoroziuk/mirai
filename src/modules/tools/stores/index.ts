export * from 'src/modules/tools/stores/tools';
export * from 'src/modules/tools/stores/groups';
