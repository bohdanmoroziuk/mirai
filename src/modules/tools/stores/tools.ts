import { uid } from 'quasar';
import { defineStore } from 'pinia';

import {
  Tool,
  ToolId,
  ToolBody,
  toolsService,
} from 'src/modules/tools';

export interface ToolsStoreState {
  tools: Tool[];
}

export const createTool = (body: ToolBody): Tool => ({
  id: uid(),
  ...body,
});

export const useToolsStore = defineStore('tools', {
  persist: true,
  state: (): ToolsStoreState => ({
    tools: [],
  }),
  actions: {
    async getTools() {
      this.tools = await toolsService.getTools();
    },
    async addTool(body: ToolBody) {
      const tool = createTool(body);

      await toolsService.addTool(tool);

      this.tools = [tool, ...this.tools];
    },
    async deleteTool(id: ToolId) {
      await toolsService.deleteTool(id);

      this.tools = this.tools.filter((tool) => tool.id !== id);
    },
  },
});
