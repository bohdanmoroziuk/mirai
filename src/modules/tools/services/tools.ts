import {
  doc,
  setDoc,
  getDocs,
  deleteDoc,
  collection,
} from 'firebase/firestore';

import { firestore } from 'src/firebase';

import { Tool, ToolId } from 'src/modules/tools';

export const TOOLS_COLLECTION_NAME = 'tools';

export const toolsService = {
  async getTools() {
    const snapshot = await getDocs(collection(firestore, TOOLS_COLLECTION_NAME));

    const tools = snapshot.docs.map((doc) => doc.data() as Tool);

    return tools;
  },
  async addTool(tool: Tool) {
    await setDoc(doc(firestore, TOOLS_COLLECTION_NAME, tool.id), tool);
  },
  async deleteTool(id: ToolId) {
    await deleteDoc(doc(firestore, TOOLS_COLLECTION_NAME, id));
  },
};
