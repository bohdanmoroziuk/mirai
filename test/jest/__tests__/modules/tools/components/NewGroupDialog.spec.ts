import { installQuasarPlugin } from '@quasar/quasar-app-extension-testing-unit-jest';
import {
  describe,
  expect,
  it,
} from '@jest/globals';
import { mount } from '@vue/test-utils';
import { QCard } from 'quasar';

import { NewGroupDialog } from 'src/modules/tools';

installQuasarPlugin();
describe('NewGroupDialog', () => {
  it('renders successfully', () => {
    const wrapper = mount(NewGroupDialog);

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('adds a new group', async () => {
    const mockGroupName = 'React';

    const wrapper = mount(NewGroupDialog);

    await wrapper.find('button').trigger('click');

    const card = wrapper.findComponent(QCard);

    await card.find('[aria-label="Name *"]').setValue(mockGroupName);

    await card.findAll('button')[1].trigger('click');

    expect(wrapper.emitted('add')).toEqual([[mockGroupName]]);
    expect(card.exists()).toBeFalsy();
  });

  it('cancels the addition of a new group', async () => {
    const wrapper = mount(NewGroupDialog);

    await wrapper.find('button').trigger('click');

    const card = wrapper.findComponent(QCard);

    await card.find('[aria-label="Name *"]').setValue('React');

    await card.findAll('button')[0].trigger('click');

    expect(wrapper.emitted('add')).toBeUndefined();
    expect(card.exists()).toBeFalsy();
  });

  it('does not fire the add event if there are validation errors', async () => {
    const wrapper = mount(NewGroupDialog);

    await wrapper.find('button').trigger('click');

    const card = wrapper.getComponent(QCard);

    await card.find('[aria-label="Name *"]').setValue('');

    await card.findAll('button')[1].trigger('click');

    expect(card.text()).toContain('Enter a value');

    expect(wrapper.emitted('add')).toBe(undefined);
  });
});
