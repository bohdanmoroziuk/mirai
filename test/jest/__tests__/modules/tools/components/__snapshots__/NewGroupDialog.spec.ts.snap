// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`NewGroupDialog renders successfully 1`] = `<button class="q-btn q-btn-item non-selectable no-outline q-btn--standard q-btn--rectangle bg-primary text-white q-btn--actionable q-focusable q-hoverable" tabindex="0" type="button"><span class="q-focus-helper"></span><span class="q-btn__content text-center col items-center q-anchor--skip justify-center row"><i class="q-icon on-left notranslate material-icons" aria-hidden="true" role="img">add</i><span class="block">Add group</span></span></button>`;
